bumble/profiles/hap.py

```diff
@@ -0,0 +1,590 @@
+        self.additional_parameters = index
+
+
+class PresetRecord:
+    '''See Hearing Access Service 2.8. Preset record.'''
+
+    class Property:
+        class Writable(OpenIntEnum):
+            CANNOT_BE_WRITTEN = 0b0
+            CAN_BE_WRITTEN = 0b1
+
+        class IsAvailable(OpenIntEnum):
+            IS_UNAVAILABLE = 0b0
+            IS_AVAILABLE = 0b1
+
+        writable: Writable
+        is_available: IsAvailable
+
+        def __bytes__(self) -> bytes:
+            return bytes([self.writable | (self.is_available << 1)])
+
+    index: int
+    properties: Property
+    name: str
+
+    def __bytes__(self) -> bytes:
+        return bytes([self.index]) + bytes(self.properties) + self.name.encode('utf-8')
+
+    def is_available(self) -> bool:
+        return (
+            self.properties.is_available
+            == PresetRecord.Property.IsAvailable.IS_AVAILABLE
+        )
+
+
+# -----------------------------------------------------------------------------
+# Server
+# -----------------------------------------------------------------------------
+class HearingAccessService(gatt.TemplateService):
+    UUID = gatt.GATT_HEARING_ACCESS_SERVICE
+
+    hearing_aid_features_characteristic: gatt.Characteristic
+    hearing_aid_preset_control_point: gatt.Characteristic
+    active_preset_index_characteristic: gatt.Characteristic
+    active_preset_index: int = 0x00
+    active_preset_index_per_device: Dict[
+        Address, int
+    ]  # TODO validate address type → public_address ?
+
+    device: Device
+
+    hearing_aid_features: HearingAidFeatures
+    preset_records: Dict[int, PresetRecord] = {}  # key is the preset index
+    read_presets_request_in_progress: bool = False
+
+    preset_changed_operations_history: List[PresetChangedOperation]
+    preset_changed_operations_per_device: Dict[Address, int]
+    preset_changed_operations_history_per_device: Dict[
+        Address, List[PresetChangedOperation]
+    ] = {}
+
+    # Keep an updated list of connected client to send notification to
+    currently_connected_clients: Set[Connection]
+
+    def __init__(
+        self, device: Device, features: HearingAidFeatures, presets: List[PresetRecord]
+    ) -> None:
+        self.device = device
+        self.hearing_aid_features = features
+        for p in presets:
+            self.preset_records[p.index] = p
+
+        with contextlib.closing(EventWatcher()) as watcher:
+
+            @watcher.on(
+                device, 'connection'
+            )  # should it be on_characteristic_subscription ?
+            async def on_connection(connection: Connection) -> None:
+                # TODO if device is bonded && device is HAP ?
+                with contextlib.closing(EventWatcher()) as watcher:
+
+                    @watcher.on(connection, 'disconnection')
+                    def on_disconnection() -> None:
+                        self.currently_connected_clients.remove(connection)
```

Comment:
I haven't see a lot of usage of watcher.  
Is this a correct usage of it ?

Should the service listen to connection ? or should it listen to both `on_connection_encryption_change` to know the bonded device and at the same time to `on_characteristic_subscription` to know the device that want to be notified with the hearing_aid_preset_control_point

Reply:
Watcher is a context helper to release listeners after some one-shot usage. Right after exiting the context, listeners will be all released. It seems your listeners need to work on a longer lifecycle so I guess they should be native listeners instead of context-managed listeners?

Reply:
You mean using calling watcher_on directly and not using the annotation ?

Reply:
Please check the suggestion. Also I suggest you to write an example like https://github.com/google/bumble/blob/main/examples/run_unicast_server.py and make it work with Android. It should be helpful to understand how asyncio and event emitter work.